x/sync/workheap.go

```diff
@@ -99,25 +91,23 @@ func (wh *workHeap) MergeInsert(item *workItem) {
 		return
 	}
 
-	var mergedBefore, mergedAfter *heapItem
-	searchItem := &heapItem{
-		workItem: &workItem{
-			start: item.start,
-		},
+	var mergedBefore, mergedAfter *workItem
+	searchItem := &workItem{
+		start: item.start,
 	}
 
 	// Find the item with the greatest start range which is less than [item.start].
 	// Note that the iterator function will run at most once, since it always returns false.
 	wh.sortedItems.DescendLessOrEqual(
 		searchItem,
-		func(beforeItem *heapItem) bool {
-			if item.localRootID == beforeItem.workItem.localRootID &&
-				maybe.Equal(item.start, beforeItem.workItem.end, bytes.Equal) {
+		func(beforeItem *workItem) bool {
+			if item.localRootID == beforeItem.localRootID &&
+				maybe.Equal(item.start, beforeItem.end, bytes.Equal) {
 				// [beforeItem.start, beforeItem.end] and [item.start, item.end] are
 				// merged into [beforeItem.start, item.end]
-				beforeItem.workItem.end = item.end
-				beforeItem.workItem.priority = math.Max(item.priority, beforeItem.workItem.priority)
-				heap.Fix(&wh.innerHeap, beforeItem.heapIndex)
+				beforeItem.end = item.end
+				beforeItem.priority = math.Max(item.priority, beforeItem.priority)
+				wh.innerHeap.Fix(wh.innerHeap.Index()[beforeItem])
```

Comment:
Since we have no duplicates I think the changes you proposed in your PR in https://github.com/ava-labs/avalanchego/pull/2146 make this a lot cleaner. I'll merge it in.